fe/endpoint_correction.py

```diff
@@ -0,0 +1,199 @@
+import functools
+
+from typing import Tuple
+import jax
+import numpy as np
+import jax.numpy as jnp
+from timemachine.potentials import bonded
+from scipy.stats import special_ortho_group
+
+def exp_u(rotation, k, beta):
+    return jnp.exp(-beta*bonded.psi(rotation, k))
+
+exp_batch = jax.jit(jax.vmap(exp_u, (0, None, None)))
+
+def sample_multiple_rotations(k, beta, size):
+    num_batches = 500
+    batch_size = 10000
+
+    samples = []
+
+    for batch_attempt in range(num_batches):
+        Rs = special_ortho_group.rvs(3, size=batch_size)
+        tests = np.random.rand(batch_size)
+        M = np.pi**2 # volume of SO(3)
+        # acceptance_prob = exp_batch(Rs, k)/M
```

Comment:
@maxentile I think we agreed to simply set M=1, but just wanted to triple check this.

Reply:
Yep! I think there was just some ambiguity about where we were putting this constant.

In rejection sampling, we need an upper bound `M` on the ratio `q_target(x) / q_proposal(x)`, so that `q_target(x) / M q_proposal(x) <= 1` everywhere.

In our case we can get a tight bound from looking at `q_target` and `q_proposal` independently.
* `q_target(x) = exp(-U(x))`, and we know `min_x U(x) = 0`, so we know `max_x q_target(x) = 1`.
* Our proposal is uniform over rotations, so `q_proposal(x)` is a constant.

If we used `normalized_q_proposal(x) = 1 / pi^2`, then `M=pi^2` would be an appropriate constant to use, guaranteeing `q_target(x) / M normalized_q_proposal(x) <= 1` everywhere.

We could just as well set `q_proposal(x) = 1` and `M=1`, which has exactly the same effect.

However, mixing and matching these could lead to errors.

For example, it would be incorrect to use `normalized_q_proposal(x) = 1 / pi^2`, `M=1`, as we could now sample `x` where `q_target(x) / M normalized_q_proposal(x) > 1`. Setting M too small like this would impact the correctness of the sampler, but would also be easy to detect by runtime assertion.

On the other hand, setting M larger than needed does not impact the correctness of the sampler, although it impacts the efficiency. I think a previous version of this code set M ~10x larger than it needed to be, by implicitly using `q_proposal(x) = 1`, and picking `M = pi^2`. Setting `M=1` here doesn't change the correctness of the sampler, but increases its acceptance rate relative to that version.

Reply:
Thanks for the detailed explanation, I'm going to document this in the code directly.